src/charm.py

```diff
@@ -138,6 +141,63 @@ def __init__(self, *args):
         self.framework.observe(self.on.config_changed, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_created, self._configure)
         self.framework.observe(self.on[PEER_RELATION_NAME].relation_changed, self._configure)
+        self.framework.observe(self.on.authorize_charm_action, self._on_authorize_charm_action)
+
+    def _on_authorize_charm_action(self, event: ActionEvent):
+        """Authorize the charm to interact with Vault."""
+        if not self.unit.is_leader():
+            event.fail("This action can only be run by the leader unit")
+            return
+        logger.info("Authorizing the charm to interact with Vault")
+        if not self._api_address:
+            event.fail("API address is not available. Something is wrong.")
+            return
+        token = event.params["token"]
+        vault = Vault(
+            url=self._api_address, ca_cert_path=self.tls.get_tls_file_path_in_charm(File.CA)
+        )
+        vault.set_token(token)
+        vault.enable_audit_device(device_type="file", path="stdout")
+        vault.enable_approle_auth()
+        vault.configure_charm_access_policy(name=CHARM_POLICY_NAME)
```

Comment:
Should we fail early if the token is incorrect (ex. doesn't have the necessary permissions)?

Reply:
This seemed reasonable at first glance, but is there any easy way to check this?

In hindsight, how can we be sure that the token has all the necessary permissions? We'd need to keep a running list of policy permissions we need access to, and update it when new requests are added. We'd also need to start reading the policy config. This seems like a lot of extra code to maintain.

Reply:
Right, I remember now, I did test this, and I figured this was decent enough:

```
$ juju run vault/leader authorize-charm token="notarealtoken"
Running operation 2 with 1 task
  - task 3 on unit-vault-0

Waiting for task 3...
Action id 3 failed: exit status 1

Uncaught Forbidden in charm code: permission denied, on post https://10.4.45.168:8200/v1/sys/audit/file
Use `juju debug-log` to see the full traceback.
ERROR the following task failed:
 - id "3" with return code 1

use 'juju show-task' to inspect the failure
```

And then in the charm debug log, we get a proper stack trace.

I'm not against catching and re-branding the error though. [I've put up a change which](https://github.com/canonical/vault-operator/pull/62/commits/20f65b585abb8ff8f99aa8b2b63c39430478c25a) looks like this, instead:

```
$ juju run vault/leader authorize-charm token="notarealtoken"
Running operation 6 with 1 task
  - task 7 on unit-vault-0

Waiting for task 7...
Action id 7 failed: Vault returned an error while authorizing the charm: permission denied, on post https://10.4.45.168:8200/v1/sys/audit/file
```